plugins/database/mongodb/mongodb.go

```diff
@@ -91,41 +80,26 @@ func (m *MongoDB) Initialize(ctx context.Context, req dbplugin.InitializeRequest
 		return dbplugin.InitializeResponse{}, fmt.Errorf("invalid username template: %w", err)
 	}
 
-	err = mapstructure.WeakDecode(req.Config, m.mongoDBConnectionProducer)
-	if err != nil {
-		return dbplugin.InitializeResponse{}, err
-	}
-
-	if len(m.ConnectionURL) == 0 {
-		return dbplugin.InitializeResponse{}, fmt.Errorf("connection_url cannot be empty-mongo fail")
-	}
-
-	writeOpts, err := m.getWriteConcern()
-	if err != nil {
-		return dbplugin.InitializeResponse{}, err
-	}
-
-	authOpts, err := m.getTLSAuth()
+	err = m.mongoDBConnectionProducer.loadConfig(req.Config)
 	if err != nil {
 		return dbplugin.InitializeResponse{}, err
 	}
 
-	m.clientOptions = options.MergeClientOptions(writeOpts, authOpts)
-
 	// Set initialized to true at this point since all fields are set,
 	// and the connection can be established at a later time.
 	m.Initialized = true
 
 	if req.VerifyConnection {
-		_, err := m.Connection(ctx)
+		client, err := m.mongoDBConnectionProducer.createClient(ctx)
```

Comment:
How come we're changing this to call createClient directly instead of having it go through `m.Connection` to do this? It looks like `m.Connection` performs all of the logic captured in `if req.VerifyConnection { ... }`.

Reply:
`Initialize` already has the mutex locked (at the top of the function),  & it will deadlock (or panic - I don't recall which) when `m.Connection()` tries to lock it again.

Reply:
I see. Do you know if we have to call `_ = c.client.Disconnect(ctx)` after the Ping call here? Asking based on the comment within `m.Connection` of wanting to re-create the session after the ping.

Reply:
Good question. We probably should if the `Ping` fails based on the comment on `Disconnect` itself to prevent any sort of resource leak. If it succeeds, we set it into the `mongoDBConnectionProducer` so it is reused the next time `Connection` is called.